ci: introduce cilium integration test

mhofstetter opened the pull request:
This commit introduces a GitHub Actions workflow which executes the Cilium integration tests.

The actual integration test performs the following steps:
* checking out Cilium repo (default: master)
* installing cilium via Cilium CLI (default: latest)
* setting up a kind cluster
* patching Cilium Agent Dockerfile with Proxy image of current branch
* building and pushing the Cilium images with the modified Envoy Proxy
* performing the Cilium connectivity tests

The following events are triggering the workflow:
* push to master
* opening pull request targeting master (https://github.com/cilium/proxy/actions/runs/4616447297/jobs/8161468972?pr=153)
* commenting on pull request with the option to override versions of Cilium & Cilium CLI (`/test-cilium-integration
  [cilium=<ciliumRepoRef>] [ciliumCli=<ciliumCliImageTag>]`)

Workflow runs triggered by a comment on the pull request are currently not linked to the actual PR via its checks because this might not be desired in every situation. But the action responds to the comment by adding the URL to the workflow run and mentioning its success/failure. Re-Triggering a failed cilium integration test is also possible via UI - e.g. that should not be the goal of this "issue comment" feature.. At best it looks like this :crossed_fingers: 

![screenshot-2023-04-04-15-29-42](https://user-images.githubusercontent.com/6699311/229808917-e1cb3261-d71e-4418-8eb7-055adbb4a1ae.png)


Fixes: https://github.com/cilium/proxy/issues/152

mhofstetter commented:
/cilium-integration-test cilium=nope ciliumCli=nope

mhofstetter commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,155 @@
+  push:
+    branches:
+      - master
+  pull_request: {}
```
replace `pull_request` with `pull_request_trigger` before merge

sayboras commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,155 @@
+  cilium-connectivity-tests:
+    timeout-minutes: 360
+    name: Cilium Connectivity Tests
+    runs-on: ubuntu-latest-64-cores-256gb
```
we can use smaller runner here (e.g. ubuntu-latest)

sayboras commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,155 @@
+          echo "PROXY_TAG=${{ github.event.pull_request.head.sha }}" >> $GITHUB_ENV
+
+      - name: Prepare variables for issue comment on a PR
+        if: github.event_name == 'issue_comment' && github.event.issue.pull_request && startsWith(github.event.comment.body, '/cilium-integration-test')
```
we are kind of familiar with /test. What do you think?

mhofstetter replied:
good point, changed it to `/test-cilium-integration`

sayboras replied:
hehe i mean just /test :D

mhofstetter replied:
oh :D then i tend to "no" because we're not triggering every test - which i would expect from a trigger comment `/test`. it's quite explicit that we're executing the cilium integration tests with explicit version of cilium and/or cilium cli.

:thinking:

sayboras commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,155 @@
+              repo: context.repo.repo,
+              comment_id: ${{ github.event.comment.id }},
+              content: '-1'
+            })
```
can we add a small step to capture sysdump in case of failure here ?

mhofstetter replied:
good point. added steps to collect and download system dump in case of failure

sayboras commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,155 @@
+            --helm-set tls.secretsBackend=k8s \
+            --agent-image=quay.io/cilium/cilium:test \
+            --operator-image=quay.io/cilium/operator-generic:test \
+            --helm-set image.pullPolicy=IfNotPresent \
+            --helm-set operator.image.pullPolicy=IfNotPresent \
```
we can just change pull policy to never

sayboras commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,155 @@
+        run: |
+          kind load docker-image \
+            --name kind \
+          quay.io/cilium/operator-generic:test \
+          quay.io/cilium/cilium:test
```
nit: indentation alignment.

sayboras commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,155 @@
+    runs-on: ubuntu-latest-64-cores-256gb
+    steps:
+      - name: Prepare variables for pushing to master
+        if: github.event_name == 'push'
```
do we need to check for base ref here i.e. branch is master ?

mhofstetter replied:
yes, because otherwise we might trigger the workflow twice when pushing to a branch with a corresponding PR (push & PR sync event)

mhofstetter replied:
@sayboras we might need to adapt this one once we branch new release branches from master

mhofstetter replied:
actually it's also covered by the branch filter in the "on push" event definition. going to remove it.

sayboras commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,155 @@
+
+      - name: Execute Cilium Connectivity Tests
+        shell: bash
+        run: cilium connectivity test
```
it's open question, do we need to enable hubble for flow validation as part of the connectivity tests ?

mhofstetter replied:
yeah, it looks like the cilium connectivity tests check some additional stuff when hubble relay is available. thanks!

sayboras reviewed (commented):
this is a huge improvement :+1:  :1st_place_medal:

mhofstetter commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,160 @@
+          echo "PROXY_TAG=${{ github.sha }}" >> $GITHUB_ENV
+
+      - name: Prepare variables for PR
+        if: github.event_name == 'pull_request'
```
i realized that we need to filter out PR's (and comments on PR's) which aren't targeting `master`. as mentioned in another comment, we need to adapt once we're branching a new release branch from `master` and want to support it there too.

mhofstetter replied:
adapted, that PR's based cilium integration tests are only triggered if the base branch of the PR is `master`. on other PR's there's still the possibility to use the variant with the issue comment on the PR.

in the future we might chose the right version of cilium & cilium-cli for other proxy release branches as well.

mhofstetter commented:
@sayboras can you please have a second look?

i addressed your comments and some other points (version pinning, define conditions on the job itself (e.g. define event & job  conditions, such that the workflow only gets triggered on pushes to master, PRs targeting master and issue comments on PRs starting with the right prefix. IMO this then allows us to drop these tests on steps of this job)

sayboras reviewed (approved):
Thanks and LGTM :100:.

Just a note that we will need to backport this change to v1.23 branch later.

jrajahalme reviewed (approved):
Thanks for this, saves a ton of time automating this!

Some nits below.

jrajahalme commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,187 @@
+        uses: helm/kind-action@d8ccf8fb623ce1bb360ae2f45f323d9d5c5e9f00 # v1.5.0
+        with:
+          version: ${{ env.KIND_VERSION }}
+          config: '.github/kind-config.yaml'
```
Use of a kind cluster with two nodes would provide more coverage.

mhofstetter replied:
@jrajahalme the config file already removes the taint from the master node to leverage both nodes.

https://github.com/cilium/cilium/blob/master/.github/kind-config.yaml#L6-L13

jrajahalme commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,187 @@
+
+      - name: Execute Cilium Connectivity Tests
+        shell: bash
+        run: cilium connectivity test
```
IMO we could limit the tests with `--test=l7` option, that would limit the test to those with `l7` in their name, which would be all we need for Envoy integration. This would shave couple of minutes off the test time.

mhofstetter replied:
Addressed with https://github.com/cilium/proxy/pull/165

jrajahalme reviewed (commented):
Some additional comments for future improvement :-)

jrajahalme commented:
Seems that run on PR push does not work, says the test is skipped: https://github.com/cilium/proxy/actions/runs/4624574857/jobs/8179587054?pr=142

(The PR in question was rebased to master, if it makes a difference)

jrajahalme commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,187 @@
+        if: github.event_name == 'issue_comment'
+        run: |
+          echo "PROXY_IMAGE=quay.io/cilium/cilium-envoy-dev" >> $GITHUB_ENV
+          echo "PROXY_TAG=${{ github.event.issue.pull_request.head.sha }}" >> $GITHUB_ENV
```
This fails on an issue comment:
```
  echo "PROXY_IMAGE=quay.io/cilium/cilium-envoy-dev" >> $GITHUB_ENV
  echo "PROXY_TAG=" >> $GITHUB_ENV
```
See the empty `PROXY_TAG`. As seen on: https://github.com/cilium/proxy/actions/runs/4624591204/jobs/8179582498

The `Wait for Cilium proxy image to be available` step never completes because of this.

mhofstetter replied:
yeah, seems like `${{ github.event.issue.pull_request.head.sha }}` is not set in case of a comment :man_facepalming:

mhofstetter replied:
addressed by https://github.com/cilium/proxy/pull/160

mhofstetter commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,187 @@
+          echo "CILIUM_REPO_REF=${ciliumRef}" >> $GITHUB_ENV
+
+          ciliumCliTag=${CILIUM_CLI_TAG}
+          if [[ "${{ github.event.comment.body }}" == *" ciliumCli="* ]]; then
```
addressed in https://github.com/cilium/proxy/pull/159

mhofstetter commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,187 @@
+
+      - name: Install Cilium CLI ${{ env.CILIUM_CLI_TAG }}
+        run: |
+          cid=$(docker create quay.io/cilium/cilium-cli-ci:${{ env.CILIUM_CLI_TAG }} ls)
```
definitely. i switched from installing the Cilium CLI via `tar.gz` download from GitHub to the Docker Image approach because the `tar.gz` are only provided for releases whereas a Docker Image is built for every build.

But agree that it's not as comfortable to provide in case of a stable release. will address this together with the other mentioned improvements from you!

mhofstetter replied:
addressed with PR https://github.com/cilium/proxy/pull/161

mhofstetter commented on .github/workflows/cilium-integration-tests.yaml:
```diff
@@ -0,0 +1,187 @@
+      - name: Checkout Cilium ${{ env.CILIUM_REPO_REF }}
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+        with:
+          repository: cilium/cilium # Be aware that this is the Cilium repository and not the one of the proxy itself!
```
@jrajahalme don't get this point. 

this is just the repository. but one has the possibility to provide the `ref` when checking out the repo (https://github.com/cilium/proxy/blob/master/.github/workflows/cilium-integration-tests.yaml#L88) - which can be anything - including your feature branch name of the CIlium Repository.

mhofstetter replied:
ah - forked repos aren't addressed - if that's what you mean.

mhofstetter replied:
addressed with https://github.com/cilium/proxy/pull/166

mhofstetter commented:
@jrajahalme addressed all your input in PR's. thanks again!
